### PR TITLE
Sink cache: crop if exceeds max cache len

### LINKED ISSUE #32233: SinkCache with Qwen1.5 broken in 4.43.0+
### System Info

- `transformers` version: 4.43.2
- Platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.35
- Python version: 3.9.16
- Huggingface_hub version: 0.24.0
- Safetensors version: 0.4.2
- Accelerate version: 0.33.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: No
- GPU type: NVIDIA RTX A6000

### Who can help?

@zucchini-nlp @gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Run the code below on CPU (using GPU hides the actual error behind a `RuntimeError: CUDA error: device-side assert triggered`).

\*\*\* **Works fine in v4.42.4, error appears only in 4.43.0+** \*\*\*

\*\*\* **Tested Llama-2, Mistral, and Qwen1.5 models. Issue only appears to affect Qwen1.5, but may impact other models that I didn't test.** \*\*\*

This appears to be a separate issue from #31381

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, SinkCache
from transformers.trainer_utils import set_seed

# Load the model and tokenizer
# model_name = "meta-llama/Llama-2-7b-hf" # <-- this works!
# model_name = "mistralai/Mistral-7B-v0.1" # <-- this works!
model_name = "Qwen/Qwen1.5-1.8B" # <-- this doesn't work!

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Prepare the input
input_text = "The quick brown fox jumps over the lazy"
inputs = tokenizer(input_text, return_tensors="pt")

# Generate the output
set_seed(42)
sink_cache = SinkCache(window_length=50, num_sink_tokens=8)
output = model.generate(
    **inputs, 
    use_cache=True, 
    past_key_values=sink_cache, 
    max_new_tokens=100,
    do_sample=True,
    top_p=0.9,
)

# Decode the output
output_text = tokenizer.decode(output[0])
print(output_text)
```

Traceback:
```text
Traceback (most recent call last):
  File "/remote/ayuser/user/sinkcache_test.py", line 19, in <module>
    output = model.generate(
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/generation/utils.py", line 1989, in generate
    result = self._sample(
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/generation/utils.py", line 2932, in _sample
    outputs = self(**model_inputs, return_dict=True)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 1054, in forward
    outputs = self.model(
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 856, in forward
    layer_outputs = decoder_layer(
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 596, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 496, in forward
    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
  File "/home/user/anaconda3/envs/bark/lib/python3.9/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 149, in apply_rotary_pos_emb
    cos = cos[position_ids].unsqueeze(unsqueeze_dim)
IndexError: index 50 is out of bounds for dimension 0 with size 50
```

### Expected behavior

The generation would complete with no error.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #32233. Previous PR completely broke SinkCache and now it is not working as supposed. This PR reverts the change regarding `max_cache_length`, because for sink cache we need to get the latest `max length` masks and have position ida not exceeding the `max length`

I am aware that SinkCache is buggy and doesn't work for some edge cases and for multi-turn dialogues. But I think we can try to keep the previous working version for one-turn generation, until we figure out how to make SinkCache happy

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Could we add a SinkCache test to make sure this does not get broken again? 
Also, SinkCache is kind of like the sliding window cache, wondering why we don;t use this:
```
        # SlidingWindowCache
        if using_sliding_window_cache:
            target_length = max(sequence_length, self.config.sliding_window)
```
which we have in mistral!

**Comment by ArthurZucker** [comment]
But the sequence length is given and we create the attention mask based off sequence_length, target_length. 
```python3
            causal_mask = torch.full(
                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
            )
            exclude_mask = torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
            if self.config.sliding_window is not None:
                if not using_sliding_window_cache or sequence_length > self.config.sliding_window:
                    exclude_mask.bitwise_or_(
                        torch.arange(target_length, device=device)
                        <= (cache_position.reshape(-1, 1) - self.config.sliding_window)
                    )
```

**Comment by ArthurZucker** [comment]
now, if slicing the mask is the only thing we need to use potentially siding window attention (or SinkCache) on models, LGTM!

**Comment by ArthurZucker** [comment]
(It means we can refactor the mistral piece of code as well)

**Comment by zucchini-nlp** [comment]
Yes, cropping right now is the only way to make SinkCache happy. I just realized moving the crop to `update_causal_mask` will cause errors in SinkCache case because we would also have to re-inii position ids. Sink Cache position ids should not go beyond max-cache-length, though I am not sure about sliding window.

I propose to leave it in `prepare_for_generation` as is. Otherwise we have to do the same checks in `prepare_for_generation` to crop `position_ids` which will not make code look any easier. Added a test and made some changes

**Comment by gante** [comment]
> Could we add a SinkCache test to make sure this does not get broken again?

We have a couple slow sink cache tests failing 🙃 I'd argue that we have so many failing slow tests that we stopped paying attention to new failures :P

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -688,6 +688,10 @@ def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
 
     def get_max_length(self) -> Optional[int]:
         """Returns the maximum sequence length of the cached states."""
+        logger.warning_once(
+            "`get_max_length` for SinkCache is deprecated and will return 0 after v4.47. ",
+            "Use past_key_values.window_length to get the maximum window length of the cache.",
+        )
```
Comment (gante) [inline src/transformers/cache_utils.py:694]: I dislike this change -- `self.window_length` is the maximum length of the cache, returning 0 (or `None`) would not be correct. I realize the use of the method is inconsistent throughout our cache classes.

We should add two new attributes to the `Cache` classes, to fully specify a cache:
- a boolean attribute to determine whether the cache grows or has a static shape (discussed in other PRs, to be used in compilation checks)
- a boolean to determine whether it uses a sliding window (to be used e.g. in this PR)

With `get_max_length` and these two attributes we could be able to define all types of cache, and control the modeling code accordingly 🤗

**Review by gante — CHANGES_REQUESTED** [review]
I think the current changes will add more fuel to the fire, rather than untangle the use of the different `Cache` classes. Added a comment below explaining my vision for this problem, for discussion 🤗

File: src/transformers/cache_utils.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/cache_utils.py:694]: Hmm, makes sense. I used the `none` following SlidingwindownCache which actually is almost same as SinkCache but does not return any max length. So we need two attributes on each cache class: `_is_sliding` and `_is_static` right? What about HybridCache, will it be sliding or not?

Comment (gante) [reply src/transformers/cache_utils.py:694]: Those two attribute names look good 👍 (btw, probably should be done in a separate PR, for better future reference)

`HybridCache` is a good corner case indeed 🤔 I think we can treat it as a NON sliding cache, as common model variables (e.g. `attention_mask`) need to be prepared according to the non-sliding cache layers.

Comment (zucchini-nlp) [reply src/transformers/cache_utils.py:694]: Got you! I will add those two attributes in a second PR. 

Just so we're on the same page, even after adding attributes we'll have to crop inputs to the max sliding window length. I guess the attr will make checks look nicer, like `if cache._is_static: length = cache.get_max_length()`

Comment (gante) [reply src/transformers/cache_utils.py:694]: Yes, on the same page as you: we need to crop things with sliding windows :)

**Comment by zucchini-nlp** [comment]
close as sink cache will not be maintained anymore

**Pull request closed without merging**